Add documentation  for multi format reader (#401)
FAIRmat-NFDI/pynxtools: docs/learn/multi-format-reader.md

READER = MyDataReader
```

In order to understand the capabilities of the `MultiFormatReader` and which methods need to be implemented when extending it, we will have a look at its ```read``` method:
```python
def read(
    self,
    template: dict = None,
    file_paths: Tuple[str] = None,
    objects: Optional[Tuple[Any]] = None,
    **kwargs,
) -> dict:
    self.kwargs = kwargs
    self.config_file = self.kwargs.get("config_file", self.config_file)
    self.overwrite_keys = self.kwargs.get("overwrite_keys", self.overwrite_keys)   
```
### Template initialization and processing order
An empty `Template` object is initialized that later gets filled from the data files later.
```python
    template = Template(overwrite_keys=self.overwrite_keys)

    def get_processing_order(path: str) -> Tuple[int, Union[str, int]]:
        """
        Returns the processing order of the file.
        """
        ext = os.path.splitext(path)[1]
        if self.processing_order is None or ext not in self.processing_order:
            return (1, ext)
        return (0, self.processing_order.index(ext))

    sorted_paths = sorted(file_paths, key=get_processing_order)
```

> If the reader has a `self.processing_order`, the input files get sorted in this order. If `self.overwrite_keys` is True, later files get precedent.

Reviewer: Maybe give an example for a processing_order entry.

Author: added

### Reading of input files
```python
    for file_path in sorted_paths:
        extension = os.path.splitext(file_path)[1].lower()
        if extension not in self.extensions:
            logger.warning(
                f"File {file_path} has an unsupported extension, ignoring file."
            )
            continue
        if not os.path.exists(file_path):
            logger.warning(f"File {file_path} does not exist, ignoring entry.")
            continue

        template.update(self.extensions.get(extension, lambda _: {})(file_path))
```
This parts reads in the data from all data files. The `MultiFormatReader` has an `extensions` property, which is a dictionary that for each file extension calls a function that reads in data from files with that extension. If the reader shall handle e.g. an HDF5 file, a method for handling this type of file should be added, i.e., `self.extensions[".hdf5"] = self.handle_hdf5`. 

> Note that these method should also implement any case selection logic, i.e., it may not be sufficient to rely on the filename suffix, but when may also need to check for multiple versions, binary signature, mimetype, etc.

Reviewer: ```suggestion
Note that these methods should also implement any logic depending on the provided data, i.e., it may not be sufficient to rely on the filename suffix, but the reader may also need to check for different file versions, binary signature, mimetype, etc.

```

Author: changed

Note that for several input formats, standardized parser functions already exist within the `MultiFormatReader`. For example, YAML files can be parsed using the `pynxtools.dataconverter.readers.utils.parse_yml` function.

### Setting default values in the template
```python
    template.update(self.setup_template())
```
Next, the `setup_template` method can be implemented, which is used to populate the template with initial data that does not come from the files themselves. This may be used to set fixed information, e.g., about the reader. As an example, `NXentry/program_name` (which is defined as the name of program used to generate the NeXus file) scan be set to `pynxtools-plugin` by making `setup_template` return a dictionary of the form
```json
{
  "/ENTRY[my_entry]/program_name": "pynxtools-plugin",
  "/ENTRY[my_entry]/program_name/@version": "v0.1.0"
}

> ```

Reviewer: Shoudn't this come before reading the files?

Author: Currently, it comes afterwards (and after potential hard-coded update of the templates in the reading functions). We can discuss to change the order if you like.

Reviewer: I don't use either option, so in general I don't mind. Let's keep as is at least for now.

### Handling objects
```python
    if objects is not None:
        template.update(self.handle_objects(objects))
```

> Aside from data file, it is also possible to directly pass any Python objects to the `read` function (e.g., a numpy array with measurement data). In order to exploit this, the `handle_objects` method must implemented, which should return a dictionary that populates the template.

Reviewer: ```suggestion
Aside from data files, it is also possible to directly pass any Python objects to the `read` function (e.g., a numpy array with measurement data). In order to exploit this, the `handle_objects` method must implemented, which should return a dictionary that populates the template.

```

Author: changed

```python
self.callbacks = ParseJsonCallbacks(
    attrs_callback=self.get_attr,
    data_callback=self.get_data,
    eln_callback=self.get_eln_data,
    dims=self.get_data_dims,
)
```
The `ParseJsonCallbacks` class has an attribute called `special_key_map` that makes use of these callbacks to populate the template based on the starting prefix of the config dict value:
```python
self.special_key_map = {
    "@attrs": attrs_callback if attrs_callback is not None else self.identity,
    "@link": link_callback if link_callback is not None else self.link_callback,
    "@data": data_callback if data_callback is not None else self.identity,
    "@eln": eln_callback if eln_callback is not None else self.identity,
}
```
That means, if the config file has an entry ```{"/ENTRY/title": "@attrs:metadata/title"}```, the `get_attr` method of the reader gets called and should return an attribute from the given path, i.e., in this case from `metadata/title`.

By default, the MultiFormatReader supports the following special prefixes:

- `@attrs`: To get metadata from the read-in experiment file(s). You need to implement the `get_attr` method in the reader.

> - `@data`: To get measurement data from the read-in experiment file(s). You need to implement the `get_data` method in the reader.

Reviewer: The destinction between data and metadata is somewhat arbitrary here. The reason to have both of these is for files/object types that require different methods to access different parts of the object, e.g. the xarrays. Maybe woth explaining this here.

Author: good point, added

- `@eln`: To get metadata from addtional ELN files. You need to implement the `get_eln_data` method in the reader.
- `@link`: To implement a link between two entities in the NeXus file. By default, the link callback returns a dict of the form {"link": value.replace("/entry/", f"/{self.entry_name}/")}, i.e., a generic `/entry/` get replaced by the actual `entry_name`.

In addition, the reader can also implement the `get_data_dims` method, which is used to return a list of the data dimensions (see below for more details).

All of `get_attr`, `get_data`, and `get_eln_data`  (as well as any similar method that might be implemented) should have the same call signature:
```python
def get_data(self, key: str, path: str) -> Any:
```

> Here, `key` is the config dict key (e.g., `"/ENTRY[my-entry]/data/data"`) and path is the path that comes _after_ the prefix in the config file. In this example, `path` would be `mydata`. With these two inputs, the reader should be able to return the correct data for this template key.

Reviewer: Not sure which example you refer to here with mydata (?)

Author: i was refering to the example config file given above. Clarified here.

### Special rules
- **Lists as config value**: It is possible to write a list of possible configurations of the sort
  ```json
  "/ENTRY/title":"['@attrs:my_title', '@eln', 'no title']"
  ```

>   The value most be a string of a list, with each item being a string itself. This allows to provide different options depending if the data exists for a given callback. For each list item , it is checked if a value can be returned and if so, the value is written. In this example, the converter would check (in order) the `@attrs` (with path `"my_title"`) and `@eln` (with path `""`) tokens and write the respective value if it exists. If not, it defaults to "no title".

Reviewer: ```suggestion
  The value must be a string which can be parsed as a list, with each item being a string itself. This allows to provide different options depending if the data exists for a given callback. For each list item , it is checked if a value can be returned and if so, the value is written. In this example, the converter would check (in order) the `@attrs` (with path `"my_title"`) and `@eln` (with path `""`) tokens and write the respective value if it exists. If not, it defaults to "no title".

```

Author: changed

>   This concept can be particularly useful if the same config file is used for multiple measurement configurations, where for some setup, the same metadata may or not be available.

Reviewer: ```suggestion
  This concept can be particularly useful if the same config file is used for multiple measurement configurations, where for some setup, the same metadata may or may not be available.

```

Author: 👍

    Note that if this notation is used, it may be helpful to pass the `suppress_warning` keyword as `True` to the read function. Otherwise, there will be a warning for every non-existent value.

- **Wildcard notation**: There exists a wildcard notation (using `*`)
  ```json

>   "/ENTRY/data/AXISNAME[*]": "@data:*.data",

Reviewer: I remember a discussion, that UPPERCASE notation is actually not valid for field names. What is the current status here? While the converter does not complain, it might just ignore it (?)

Author: This can only for the `NXdata` where we have `AXISNAME` and `DATA` and there is some more logic (like `signal`, `axes` and so on) to distinguish the two. The verification handles this special case differently. Other than that, it is indeed not allowed.

  ```
  that allows filling multiple fields of the same type from a list of dimensions. This can be particularly helpful for writing `DATA` and `AXISNAME` fields that are all stored under similar paths in the read-in data.

>   For this, the `get_data_dims` method need to implemented. For a given path, it should return a list of all data axes available to replace the wildcard.

Reviewer: ```suggestion
  For this, the `get_data_dims` method needs to be implemented. For a given path, it should return a list of all data axes available to replace the wildcard.

```

Author: changed

>     The same wildcard notation can also be used with a name to repeat entries with different names (e.g., *{my, name, etc} replaces to three keys with my, name, etc replaced, respectively). As an example, for multiple lenses and their voltage readouts, one could write:

Reviewer: ```suggestion
    The same wildcard notation can also be used within a name to repeat entries with different names (e.g., field_*{my, name, etc} is converted into three keys with * replaced by my, name, etc, respectively). As an example, for multiple lenses and their voltage readouts, one could write:

```

Author: that's clearer, thanks

  ```json
  "LENS_EM[lens_*{A,B,C,Foc}]": {
    "name": "*",
    "voltage": "@attrs:metadata/file/Lens:*:V",
    "voltage/@units": "V"
  },
  ```

>   which would write `NXlens_em` instances named `lens_A`, `lens_B`, and `lens_Foc`.

Reviewer: ```suggestion
  which would write `NXlens_em` instances named `lens_A`, `lens_B`, `lens_C`, and `lens_Foc`.

```

Author: just removed the C from the config to make it shorter

>     Note that the `get_data_dims` must correspond to all use cases, i.e., for `@data`, `@attrs`, and so on.

Reviewer: This I don't understand. Will this not just replace "*" by the entries in the dict? Does this use case really require get_data_dims?

Author: removed

- **Required fields in optional groups**: There will sometimes be the situation that there is an optional NeXus group in an application definition, that (if implemented) requires some sub-element. As an example, for the instrument's energy resolution, the only value expected to come from a data source is the `resolution`, whereas other fields are hardcoded.
  ```json
  "ENTRY/INSTRUMENT[instrument]/energy_resolution": {
    "resolution": "@attrs:metadata/instrument/electronanalyser/energy_resolution",
    "resolution/@units": "meV",
    "physical_quantity": "energy"
  }
  ```

>   Now, if there is no data for `@attrs:metadata/instrument/electronanalyser/energy_resolution` available in a dataset, this will be skipped by the reader, and not available, yet the other entries are present. During validation, this means that the required field resolution of the optional group `energy_resolution` is not present, and thus an error would be raised:

Reviewer: ```suggestion
  Now, if there is no data for `@attrs:metadata/instrument/electronanalyser/energy_resolution` available in a dataset, this will be skipped by the reader, and not available, yet the other entries are present. During validation, this means that the required field `resolution` of the optional group `energy_resolution` is not present, and thus a warning or error would be raised:

```

Author: correct
